### PR TITLE
AIRFLOW-21 upgrade GCP client lib

### PR DESCRIPTION
**Pull request by alexvanboxel**
Dear Airflow Maintainers,

Please accept the following PR that
- Addresses the following issues (links to issues addressed by this PR) :
  AIRFLOW-21

### CODE REVIEW

**Comment by bolkedebruin** [comment]
@criccomini

**Comment by jlowin** [comment]
Can confirm this works wonderfully without a keyfile (using `gcloud auth`). Still have this one odd message in the logs but I don't know if it's on the Airflow side or not: 

```
INFO:root:Getting connection using `gcloud auth` user, since no key file is defined for hook.
WARNING:oauth2client.util:__init__() takes at most 2 positional arguments (3 given)
```

In any case, it works fine for reading/writing logs to GCS with the newest versions of the libraries.

+1 from me but @criccomini should have a look before merging

**Comment by alexvanboxel** [comment]
@jlowin did you update the dependencies before testing? I've updated setup.py and it should update the libs. If it's still the case I can have a look at it (but not today...)

**Comment by jlowin** [comment]
Yup, I installed `airflow[gcp_api]` and confirmed the new versions were installed.

File: airflow/contrib/hooks/gcp_api_base_hook.py [hunk 1]
```diff
@@ -29,6 +32,16 @@ class GoogleCloudBaseHook(BaseHook):
     for a Google cloud service.
 
     The class also contains some miscellaneous helper functions.
+
+    All hook derived from this base hook use the 'Google Cloud Platform' connection
+    type. Two ways of authentication are supported:
+
+    Default credentials: Only specify 'Project Id'. Then you need to have executed
```
Comment (criccomini) [inline airflow/contrib/hooks/gcp_api_base_hook.py:39]: Does scope need to be supplied when using default credentials? I haven't tried this.

File: airflow/contrib/hooks/gcp_api_base_hook.py [hunk 2]
```diff
@@ -48,27 +61,27 @@ def _authorize(self):
...
-                credentials = SignedJwtAssertionCredentials(
-                    service_account,
-                    key,
-                    scope=self.scope,
-                    **kwargs)
         else:
-            raise AirflowException('Scope undefined, or either key_path/service_account config was missing.')
+            if not scope:
+                raise AirflowException('Scope should be defined when using a key file.')
+            scopes = scope.split(',')
```
Comment (criccomini) [inline airflow/contrib/hooks/gcp_api_base_hook.py]: Wonder if we should:

```
[trim(s) for s in scope]
```

To handle extra white space (e.g. `http://foo, http:/bar`)

File: airflow/contrib/hooks/gcp_api_base_hook.py [hunk 3]
```diff
@@ -85,3 +98,7 @@ def _get_field(self, f, default=None):
             return self.extras[long_f]
         else:
             return default
+
+    @property
+    def project_id(self):
```
Comment (criccomini) [inline airflow/contrib/hooks/gcp_api_base_hook.py:107]: Is this used anywhere?

File: airflow/contrib/hooks/gcp_api_base_hook.py (hunk 2 above)
Comment (alexvanboxel) [inline airflow/contrib/hooks/gcp_api_base_hook.py]: will change

File: airflow/contrib/hooks/gcp_api_base_hook.py (hunk 3 above)
Comment (alexvanboxel) [inline airflow/contrib/hooks/gcp_api_base_hook.py:107]: yes, I need it in my DataFlow/DataProc hook

**Pull request merged**